### ISSUE TITLE
garp should be send when 1 of the keepalived servers stops being master

### ISSUE THREAD

**Issue by dupondje** [acassen/keepalived#542]
Hi,

We have some (edge) case where keepalived VIP goes down. And manual actions are needed.

Situation:
2 servers with keepalived to keep 1 virtual IP active on 1 of both machines.
server1 => Master
server2 => Backup

Case:
Both servers are being backed up by Veeam (ESXi).

Now ESXi freezes the machine for some moments (mostly 1-2 seconds) to commit its snapshot after the backup.
This shouldn't give a problem, but this isn't the case.

What happens:
- server1 is being backed up
- snapshot on server1 is commited, and server1 is freezed for 1-2 seconds
- server2 notices this, and becomes MASTER
Mar 16 03:03:50 server2 Keepalived_vrrp[580]: VRRP_Instance(VIP1) Transition to MASTER STATE
Mar 16 03:03:51 server2 Keepalived_vrrp[580]: VRRP_Instance(VIP1) Entering MASTER STATE
- snapshot is commited, and server1 is back
- server2 notices this, and enters backup state again
Mar 16 03:03:51 server2 Keepalived_vrrp[580]: VRRP_Instance(VIP1) Received higher prio advert 250
Mar 16 03:03:51 server2 Keepalived_vrrp[580]: VRRP_Instance(VIP1) Entering BACKUP STATE

But the big issue here is, server1 didn't know some other server was MASTER for a seconds (and did send a garp).
So server1 doesn't send a garp anymore, and the VIP is down because of this.

Solutions?
- We could enable vrrp_garp_master_refresh, but this will cause a lot of additional (useless) garp's on our network
- The server2 should send some inform to server1: 'hey I was master, please garp your VIP's again.

**Comment by pqarmitage** [comment]
I agree enabling vrrp_garp_master_refresh is not ideal. Re server 2 sending a message to server1, there is no element of the VRRP protocol that does this.

I presume you don't want to extend your advert_int to stop server 2 timing out and becoming master. If the freeze is only for 1-2 seconds, then the advert interval must be very low (? < 0.5 seconds). Extending the advert_int to be half the maximum freeze time should stop server 2 taking over as master.

The best idea that I can think of is to force server 1 to become backup before the snapshot is committed. You could use a non-weighted track script, that returns a non-zero exit code prior to the snapshot being taken. This will cause the server 1 VRRP instance to go to fault state, it should send a priority 0 advert and so server 2 will take over as master. The lower the value of `fall`, and the lower the `interval` value configured against the script, the sooner the VRRP instance will go to fault state. server 2 should then take over as master within 1 advert interval (the higher its priority the faster it will take over).

Once the snapshot is complete, the track script should start returning a 0 exit code, and so server 1 will transition to backup and then master state again.

**Comment by dupondje** [comment]
Thanks for your quick response.

The advert_int is just default (1).
We do daily backups, but it only happens like once in 30 days that this happens. Might be a timeout of 4 seconds also in such a case. So extending advert_int could help indeed.

What is also strange is that vrrp_garp_master_delay is set to 5 (default), so that means timeout was longer then 7seconds? (5 seconds for vrrp_garp_master_delay and 2s for advert_int*2)

We could add some logic in the backups indeed. But I can think of other cases where this can happen also.
For example if the link between both machines is down, but they can still reach gateway, so VIP is up. Both machines will become MASTER. When link is restored, server2 will become backup again, but server1 will not garp again.

**Comment by pqarmitage** [comment]
The vrrp_garp_master_delay has no impact on the timeout for a master to transition to backup. The timeout depends on whether you are running VRRPv2 or VRRPv3. 

For VRRPv2, the timeout is 3 * advert interval + (256 - priority) / 256 seconds.
For VRRPv3, the timeout is 3 * advert interval + (256 - priority) / 256 * advert interval seconds.

So in your case, with an advert interval of 1 second, and a priority on the master of 250, the timeout is 3.02 seconds. On the backup, the timeout will be somewhere between 3.02 and 4 seconds, depending on the priority.

The vrrp_garp_master_delay is the delay between sending the first batch of GARPs after transitioning to master, and the second batch.

The only way I can think of to resolve the dual master issue and garps is if a master receives a higher priority advert, it sends an advert itself before transitioning to backup. If a master receives a lower priority advert it should send more garps (this is configured via garp_lower_prio_repeat). Sending an advert after receiving a higher priority advert does not conform to the RFCs, but could be implemented as a non-standard option.

**Comment by pqarmitage** [comment]
Commit 72fd270 (and 5595e5e) adds the configuration options `vrrp_higher_prio_send_advert` as a global option, and `higher_prio_send_advert` as a per vrrp instance option, that implement what I described above.

So if with the new code you set `higher_prio_send_advert` on each vrrp instance, and also have `garp_lower_prio_repeat 1` (the value is your choice for how many GARPs to send), when a lower priority master receives a higher priority advert, it will send a (lower priority) advert before transitioning to backup, and the higher priority master, when it receives the lower priority advert, will send GARP messages.

This should resolve your problem, and also the problem you identified if the link goes down between the 2 machines.

**Comment by senthilnathann** [comment]
Hi,

vrrp_higher_prio_send_advert is not working in our keepalived configurations.

Our keepalived version 1.3.9

Our MASTER config:

global_defs {
 ...
vrrp_higher_prio_send_advert true
vrrp_garp_lower_prio_repeat 2
vrrp_garp_lower_prio_delay 30
}
BACKUP Config:

global_defs {
...
vrrp_garp_lower_prio_repeat 2
vrrp_garp_lower_prio_delay 30
}

Logs from backup:

Keepalived_vrrp[21821]: VRRP_Instance(VI_GATEWAY) Received advert with higher priority 100, ours 100
VRRP_Instance(VI_GATEWAY) Entering BACKUP STATE
VRRP_Instance(VI_GATEWAY) removing protocol VIPs
VRRP_Group(VG1) Syncing instances to BACKUP state
VRRP_Instance(VI_1) Entering BACKUP STATE

So no low periority messages sent to MASTER and MASTER not sent any garp messages .

Kindly check and give me a solution on this.

**Comment by pqarmitage** [comment]
@senthilnathann Can you please provide full copies of the configurations for MASTER and BACKUP, and also the full keepalived logs of both MASTER and BACKUP. We can then investigate this issue.

**Comment by senthilnathann** [comment]
Thanks for your quick reply,actually due to network disturbance ,we are faced split brain issue.After issue solved ,we need to update to GARP messages from current MASTER,so we tried this solution ,but its not working .Pls find the below config details .

MASTER Config :

global_defs {
        notification_email {
                XXXXX
        }
        notification_email_from XXXXX
        smtp_server XXXXXX
        smtp_connect_timeout 30
        router_id LVSID01
        vrrp_higher_prio_send_advert 
        vrrp_garp_lower_prio_repeat 2
        vrrp_garp_lower_prio_delay 30
}
vrrp_sync_group VG1 {
        group {
                VI_1
                VI_GATEWAY
        }
}

vrrp_instance VI_GATEWAY {
        state EQUAL
        interface bond0
        lvs_sync_daemon_inteface bond0
        virtual_router_id 98
        priority 100
        advert_int 1
        smtp_alert
        authentication {
                auth_type PASS
                auth_pass XXXX
        }
        virtual_ipaddress {
                XXXX
        }
}
vrrp_instance VI_1 {
        state EQUAL
        interface bond0
        lvs_sync_daemon_inteface bond0
        virtual_router_id 99
        priority 100
        advert_int 1
        smtp_alert
        authentication {
                auth_type PASS
                auth_pass XXXX
        }
        virtual_ipaddress {
                172.20.31.90

        }
}

BACKUP:

global_defs {
        notification_email {
                XXXXX
        }
        notification_email_from XXXXX
        smtp_server 192.168.100.16
        smtp_connect_timeout 30
        router_id LVSID01
        vrrp_garp_lower_prio_repeat 2
        vrrp_garp_lower_prio_delay 30
}
vrrp_sync_group VG1 {
        group {
                VI_1
                VI_GATEWAY
        }
}

vrrp_instance VI_GATEWAY {
        state EQUAL
        interface bond0
        lvs_sync_daemon_inteface bond0
        virtual_router_id 98
        priority 100
        advert_int 1
        smtp_alert
        authentication {
                auth_type PASS
                auth_pass XXXXX
        }
        virtual_ipaddress {
                172.20.131.115
        }
}
vrrp_instance VI_1 {
        state EQUAL
        interface bond0
        lvs_sync_daemon_inteface bond0
        virtual_router_id 99
        priority 100
        advert_int 1
        smtp_alert
        authentication {
                auth_type PASS
                auth_pass XXXXX
        }
        virtual_ipaddress {
                172.20.31.90

        }
}

> MASTER :
> 
> Sep 17 21:51:44 l3-newversion-1 Keepalived_vrrp[6836]: VRRP_Instance(VI_GATEWAY) Received advert with higher priority 100, ours 100
> Sep 17 21:51:44 l3-newversion-1 Keepalived_vrrp[6836]: VRRP_Instance(VI_GATEWAY) Entering BACKUP STATE
> Sep 17 21:51:44 l3-newversion-1 Keepalived_vrrp[6836]: VRRP_Instance(VI_GATEWAY) removing protocol VIPs.
> Sep 17 21:51:44 l3-newversion-1 Keepalived_vrrp[6836]: VRRP_Group(VG1) Syncing instances to BACKUP state
> Sep 17 21:51:44 l3-newversion-1 Keepalived_vrrp[6836]: VRRP_Instance(VI_1) Entering BACKUP STATE
> Sep 17 21:51:44 l3-newversion-1 Keepalived_vrrp[6836]: VRRP_Instance(VI_1) removing protocol VIPs.
> 
> BACKUP :
> Sep 17 21:51:13 l3-newversion-2 Keepalived_vrrp[30437]: VRRP_Instance(VI_1) Transition to MASTER STATE
> Sep 17 21:51:13 l3-newversion-2 Keepalived_vrrp[30437]: VRRP_Group(VG1) Syncing instances to MASTER state
> Sep 17 21:51:13 l3-newversion-2 Keepalived_vrrp[30437]: VRRP_Instance(VI_GATEWAY) Transition to MASTER STATE
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: VRRP_Instance(VI_GATEWAY) Entering MASTER STATE
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: VRRP_Instance(VI_GATEWAY) setting protocol VIPs.
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: VRRP_Instance(VI_GATEWAY) Sending/queueing gratuitous ARPs on bond0 for 172.20.131.115
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: Sending gratuitous ARP on bond0 for 172.20.131.115
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: Remote SMTP server [192.168.100.16]:25 connected.
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: VRRP_Instance(VI_1) Entering MASTER STATE
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: VRRP_Instance(VI_1) setting protocol VIPs.
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: Sending gratuitous ARP on bond0 for 172.20.31.90
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: VRRP_Instance(VI_1) Sending/queueing gratuitous ARPs on bond0 for 172.20.31.90
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: Sending gratuitous ARP on bond0 for 172.20.31.90
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: Sending gratuitous ARP on bond0 for 172.20.31.90
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: Sending gratuitous ARP on bond0 for 172.20.31.90
> Sep 17 21:51:14 l3-newversion-2 Keepalived_vrrp[30437]: Sending gratuitous ARP on bond0 for 172.20.31.90
> 
>

**Comment by pqarmitage** [comment]
First, a couple of comments on your configuration:
- `lvs_sync_daemon_inteface` should be `lvs_sync_daemon_interface`.
- 'lvs_sync_daemon_interface' should only be specified against one vrrp instance. Better still, use `lvs_sync_daemon` in `global_defs`.
- There is no `state EQUAL` - just omit the line altogether

I'm a bit confused about what you refer to as MASTER and BACKUP, since it appears that BACKUP is higher priority than MASTER, (they have the same configured priority but the IP address of BACKUP must be higher that MASTER's IP address) since when they can communicate with each other, what you refer to as BACKUP becomes master and vice versa.

As you state, the use of `vrrp_higher_prio_send_advert` and `vrrp_garp_lower_prio_repeat` are designed for the recovery from a split brain situation, and enable the higher priority keepalived instance to send GARP messages when the lower priority instance transitions to backup. This functionality is basically working, but following investigating what you have reported I can see there is a problem when using sync groups.

The problem is that with a sync group, when recovering from a split brain situation, when the first instance sees a higher priority advert, it sends a lower priority advert, transitions to backup and makes the other instances in the sync group transition to backup. The problem is that those instances haven't seen a higher priority advert and hence don't send another (lower priority) advert _[remember that this functionality is a non-standard, but compatible, extension to the vrrp protocol that I created]_.

I need to think about the best way to deal with this problem but at the moment I think what should probably happen is that when a member of a sync group sends GARP messages due to `garp_lower_priority_repeat` then the other members of the sync group that have `garp_lower_priority_repeat` set should also send GARP messages.

**Comment by pqarmitage** [comment]
@senthilnathann Commit da3fd4b implements the solution outlined above. In order to incorporate the commit you will need to upgrade to the latest HEAD from git, which is quite an upgrade from v1.3.9, although keepalived should be fully backwards compatible.

You could, if you want, backport the patch to v1.3.9, although there are very many advantages in upgrading to the latest code including bug fixes, enhancements, efficiency improvements and scalability.

If this commit doesn't resolve your issues, please open a new issue and include full details (configurations, all log entries from when keepalived starts etc).
